src/gt4py/next/iterator/transforms/inline_lifts.py

```diff
@@ -139,125 +164,208 @@ class Flag(enum.IntEnum):
-                args = node.args[0].args
-                if len(args) == 0:
-                    return ir.Literal(value="True", type="bool")
-
-                res = ir.FunCall(fun=ir.SymRef(id="can_deref"), args=[args[0]])
-                for arg in args[1:]:
-                    res = ir.FunCall(
-                        fun=ir.SymRef(id="and_"),
-                        args=[res, ir.FunCall(fun=ir.SymRef(id="can_deref"), args=[arg])],
-                    )
-                return res
-        elif (
-            self.flags & self.Flag.INLINE_LIFTED_ARGS
-            and _is_lift(node)
-            and len(node.args) > 0
-            and self.predicate(node, is_scan_pass_context)
+            return res
+        return None
+
+    def transform_inline_lifted_args(self, node: ir.FunCall, **kwargs):  # noqa: C901
+        symtable, is_scan_pass_context = kwargs["symtable"], kwargs.get(
+            "is_scan_pass_context", False
+        )
+        if not (
+            _is_lift(node) and len(node.args) > 0 and self.predicate(node, is_scan_pass_context)
         ):
-            stencil = node.fun.args[0]  # type: ignore[attr-defined] # node already asserted to be of type ir.FunCall
-            eligible_lifted_args = [
-                _is_lift(arg) and self.predicate(arg, is_scan_pass_context) for arg in node.args
-            ]
-
-            if isinstance(stencil, ir.Lambda) and any(eligible_lifted_args):
-                # TODO(tehrengruber): we currently only inlining opcount preserving, but what we
-                #  actually want is to inline whenever the argument is not shifted. This is
-                #  currently beyond the capabilities of the inliner and the shift tracer.
-                new_arg_exprs: dict[ir.Sym, ir.Expr] = {}
-                inlined_args = []
-                for i, (arg, eligible) in enumerate(zip(node.args, eligible_lifted_args)):
-                    if eligible:
-                        assert isinstance(arg, ir.FunCall)
-                        inlined_arg, _ = _transform_and_extract_lift_args(
-                            arg, symtable, new_arg_exprs
+            return None
+
+        stencil = node.fun.args[0]  # type: ignore[attr-defined] # node already asserted to be of type ir.FunCall
+
+        if not isinstance(stencil, ir.Lambda):
+            return None
+
+        eligible_lifted_args = [False] * len(node.args)
+        for i, (param, arg) in enumerate(zip(stencil.params, node.args, strict=True)):
+            if not _is_lift(arg):
+                continue
+            # we don't want to inline lift args derefed at multiple positions as this would
+            # disallow creating a temporary for them if the (outer) lift can not be extracted into
+            # a temporary.
+            if self.inline_single_pos_deref_lift_args_only:
+                if param.annex.recorded_shifts not in [set(), {()}]:
+                    continue
+            eligible_lifted_args[i] = True
```

Comment:
extract to function

Reply:
Good idea, this used to be a simple one-liner.